From d3b2056c39eb59d4a74f9aac1c4bbd7d64bf71f4 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ball=C3=B3=20Gy=C3=B6rgy?= <ballogyor@gmail.com>
Date: Thu, 23 Oct 2025 02:49:46 +0200
Subject: [PATCH] Check if running under X11 on launch

The xrandr tool relies on X11, and cannot be used under Wayland to set
the resolution. Just show a message dialog and exit in that case.
---
 src/lxrandr.c | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/src/lxrandr.c b/src/lxrandr.c
index 8373e6b..53ebb4e 100644
--- a/src/lxrandr.c
+++ b/src/lxrandr.c
@@ -28,6 +28,7 @@
 
 #include <gtk/gtk.h>
 #include <glib/gi18n.h>
+#include <gdk/gdkx.h>
 
 #include <locale.h>
 #include <stdio.h>
@@ -772,6 +773,24 @@ int main(int argc, char** argv)
 
     gtk_init( &argc, &argv );
 
+#if GTK_CHECK_VERSION(3, 0, 0)
+    /* check if not running under X11 */
+    if (!GDK_IS_X11_DISPLAY(gdk_display_get_default()))
+    {
+        GtkWidget *dialog = gtk_message_dialog_new(
+            NULL,
+            GTK_DIALOG_MODAL,
+            GTK_MESSAGE_ERROR,
+            GTK_BUTTONS_OK,
+            _("This tool can be used only under X11.")
+        );
+
+        gtk_dialog_run(GTK_DIALOG(dialog));
+        gtk_widget_destroy(dialog);
+        return 1;
+    }
+#endif
+
     if( ! get_xrandr_info() )
     {
         dlg = gtk_message_dialog_new( NULL,
